File: Kotlin/src/K12.kt
import java.util.*
import kotlin.math.sqrt
import kotlin.random.Random

data class Person(var Name: String = "", var BirthYear: Int = 0, var Height: Float = 0.0f, var Weight: Float = 0.0f)
data class Staff(val Name: String, val Gender: Char, val Age: Int)
data class Position(var X: Int, var Y: Int, var Distance: Double = 0.0)

class K12 : IExecutable() {
    override fun question_01() {
        val person = Person()
        print("名前を入力してください＞")
        val reader = Scanner(System.`in`)
        person.Name = reader.nextLine()
        print("生年（西暦）を入力してください＞")
        person.BirthYear = reader.nextInt()
        print("身長を入力してください＞")
        person.Height = reader.nextFloat()
        print("体重を入力してください＞")
        person.Weight = reader.nextFloat()

        println("プロファイル")
        println("名前：${person.Name}")
        println("生年：${person.BirthYear}")
        println("身長：${person.Height}")
        println("体重：${person.Weight}")
    }

    override fun question_02() {
        val staffs = arrayOf(
            Staff("神木隆之介", 'M', 23),
            Staff("上白石萌音", 'F', 18),
            Staff("長澤まさみ", 'F', 28)
            )

        println("名前\t\t性別\t年齢")
        for (x in 1..35) {
            print('-')
        }
        println()

        for (staff in staffs) {
            println("${staff.Name}\t${staff.Gender}\t${staff.Age}歳")
        }
    }

    fun getDistance(playerPos: Position, enemyPos: Position): Double {
        val diffX = playerPos.X - enemyPos.X
        val diffY = playerPos.Y - enemyPos.Y
        return sqrt((diffX * diffX + diffY * diffY).toDouble())
    }

    override fun question_03() {
        val getRandomNumber: () -> Int = { Random.nextInt(100) - 50 }
        val playerPos = Position(getRandomNumber(), getRandomNumber())
        val enemyPos = arrayOfNulls<Position>(5)
        for (x in 0..4) {
            enemyPos[x] = Position(
                getRandomNumber(),
                getRandomNumber()
            )
            enemyPos[x]!!.Distance = getDistance(playerPos, enemyPos[x]!!)
        }

        println("プレーヤーの座標：x= ${playerPos.X}\t\ty= ${playerPos.Y}")
        for ((i, pos) in enemyPos.requireNoNulls().withIndex()) {
            val str = String.format("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f",
                i, pos.X.toFloat(), pos.Y.toFloat(), pos.Distance)
            println(str)
        }

    }

    override fun question_04() {
        val getRandomNumber: () -> Int = { Random.nextInt(100) - 50 }
        val playerPos = Position(getRandomNumber(), getRandomNumber())
        val enemyPos = arrayOfNulls<Position>(5)
        for (x in 0..4) {
            enemyPos[x] = Position(
                getRandomNumber(),
                getRandomNumber()
            )
            enemyPos[x]!!.Distance = getDistance(playerPos, enemyPos[x]!!)
        }

        println("プレーヤーの座標：x= ${playerPos.X}\t\ty= ${playerPos.Y}")
        for ((i, pos) in enemyPos.requireNoNulls().withIndex()) {
            val str = String.format("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f",
                i, pos.X.toFloat(), pos.Y.toFloat(), pos.Distance)
            println(str)
        }

        println("\n並び替え後")
        enemyPos.requireNoNulls().sortBy {
            it.Distance
        }
        for ((i, pos) in enemyPos.requireNoNulls().withIndex()) {
            val str = String.format("敵キャラの座標0%d：x= %5.2f\t\ty= %5.2f\t\t距離= %5.2f",
                i, pos.X.toFloat(), pos.Y.toFloat(), pos.Distance)
            println(str)
        }
    }

}

File: CSharp/K12.cs
using System;
using System.Linq;

namespace CSharp
{
    class K12 : IExecutable
    {
        internal struct Person
        {
            internal string Name;
            internal int BirthYear;
            internal float Height;
            internal float Weight;
        }

        internal struct Staff
        {
            internal string Name;
            internal char Gender;
            internal int Age;

            internal Staff(string name, char gender, int age)
            {
                Name = name;
                Gender = gender;
                Age = age;
            }
        }

        internal struct Position : IComparable<Position>
        {
            internal int X;
            internal int Y;
            internal double Distance;

            internal Position(int x, int y, double distance = 0.0)
            {
                X = x;
                Y = y;
                Distance = distance;
            }


            public int CompareTo(Position other)
            {
                return Distance.CompareTo(other.Distance);
            }
        }

        public override void Question01()
        {
            Person person = new Person();
            Console.Write("名前を入力してください＞");
            person.Name = Console.ReadLine();
            Console.Write("生年（西暦）を入力してください＞");
            person.BirthYear = int.Parse(Console.ReadLine());
            Console.Write("身長を入力してください＞");
            person.Height = float.Parse(Console.ReadLine());
            Console.Write("体重を入力してください＞");
            person.Weight = float.Parse(Console.ReadLine());

            Console.WriteLine("プロファイル：");
            Console.WriteLine($"名前：{person.Name}");
            Console.WriteLine($"生年：{person.BirthYear}");
            Console.WriteLine($"身長：{person.Height}");
            Console.WriteLine($"体重：{person.Weight}");
        }

        public override void Question02()
        {
            Staff[] staffs =
            {
                new Staff("神木隆之介", 'M', 23),
                new Staff("上白石萌音", 'F', 18),
                new Staff("長澤まさみ", 'F', 28)
            };

            Console.WriteLine("名前\t\t性別\t年齢");
            for (int i = 0; i <= 35; ++i)
            {
                Console.Write('-');
            }
            Console.WriteLine();

            foreach (var staff in staffs)
            {
                Console.WriteLine($"{staff.Name}\t{staff.Gender}\t{staff.Age}歳");
            }

        }

        public override void Question03()
        {
            Random random = new Random();
            Func<int> getRandomNumber = () => random.Next(0, 101) - 50;
            Position playerPos = new Position(getRandomNumber(), getRandomNumber());
            Position[] enemyPos = new Position[5];
            for (int i = 0; i < enemyPos.Length; i++)
            {
                enemyPos[i].X = getRandomNumber();
                enemyPos[i].Y = getRandomNumber();
                enemyPos[i].Distance = GetDistance(playerPos, enemyPos[i]);
            }

            Console.WriteLine($"プレーヤーの座標：x= {playerPos.X}\t\ty= {playerPos.Y}\n");
            for (int i = 0; i < enemyPos.Length; i++)
            {
                var pos = enemyPos[i];
                var str = $"敵キャラの座標0{i}：x= {pos.X,5:F2}\t\ty= {pos.Y,5:F2}\t\t距離={Math.Round(pos.Distance, 2)}";
                Console.WriteLine(str);
            }

        }

        public override void Question04()
        {
            Random random = new Random();
            Func<int> getRandomNumber = () => random.Next(0, 101) - 50;
            Position playerPos = new Position(getRandomNumber(), getRandomNumber());
            Position[] enemyPos = new Position[5];
            for (int i = 0; i < enemyPos.Length; i++)
            {
                enemyPos[i].X = getRandomNumber();
                enemyPos[i].Y = getRandomNumber();
                enemyPos[i].Distance = GetDistance(playerPos, enemyPos[i]);
            }

            Console.WriteLine($"プレーヤーの座標：x= {playerPos.X}\t\ty= {playerPos.Y}\n");
            for (int i = 0; i < enemyPos.Length; i++)
            {
                var pos = enemyPos[i];
                Console.WriteLine($"敵キャラの座標0%d：x= {pos.X, 5:F2}\t\ty= {pos.Y, 5:F2}\t\t距離={Math.Round(pos.Distance, 2)}");
            }

            Console.WriteLine("\n並び替え後");
            var enemyList = enemyPos.ToList();
            enemyList.Sort();
            for (int i = 0; i < enemyList.Count; i++)
            {
                var pos = enemyList[i];
                Console.WriteLine($"敵キャラの座標0%d：x= {pos.X, 5:F2}\t\ty= {pos.Y, 5:F2}\t\t距離={Math.Round(pos.Distance, 2)}");
            }
        }

        private double GetDistance(Position playerPos, Position enemyPos)
        {
            int diffX = playerPos.X - enemyPos.X;
            int diffY = playerPos.Y - enemyPos.Y;
            return Math.Sqrt(diffX * diffX + diffY * diffY);
        }
    }
}
